DynamicModeler/Logic/vtkSlicerDynamicModelerRevolveTool.cxx

```diff
@@ -0,0 +1,375 @@
+vtkToolNewMacro(vtkSlicerDynamicModelerRevolveTool);
+
+const char* REVOLVE_INPUT_MODEL_REFERENCE_ROLE = "Revolve.InputModel";
+const char* REVOLVE_INPUT_MARKUPS_REFERENCE_ROLE = "Revolve.InputMarkups";
+const char* REVOLVE_OUTPUT_MODEL_REFERENCE_ROLE = "Revolve.OutputModel";
+const char* REVOLVE_ANGLE_DEGREES = "Revolve.AngleDegrees";
+const char* REVOLVE_AXIS_AT_ORIGIN = "Revolve.AxisAtOrigin";
+const char* REVOLVE_TRANSLATE_ALONG_AXIS_DISTANCE = "Revolve.TranslateDistanceAlongAxis";
+
+//----------------------------------------------------------------------------
+vtkSlicerDynamicModelerRevolveTool::vtkSlicerDynamicModelerRevolveTool()
+{
+  /////////
+  // Inputs
+  vtkNew<vtkIntArray> inputModelEvents;
+  inputModelEvents->InsertNextTuple1(vtkCommand::ModifiedEvent);
+  inputModelEvents->InsertNextTuple1(vtkMRMLModelNode::MeshModifiedEvent);
+  inputModelEvents->InsertNextTuple1(vtkMRMLTransformableNode::TransformModifiedEvent);
+  vtkNew<vtkStringArray> inputModelClassNames;
+  inputModelClassNames->InsertNextValue("vtkMRMLModelNode");
+  NodeInfo inputModel(
+    "Model",
+    "Model to be revolved.",
+    inputModelClassNames,
+    REVOLVE_INPUT_MODEL_REFERENCE_ROLE,
+    true,
+    false,
+    inputModelEvents
+  );
+  this->InputNodeInfo.push_back(inputModel);
+
+  vtkNew<vtkIntArray> inputMarkupEvents;
+  inputMarkupEvents->InsertNextTuple1(vtkCommand::ModifiedEvent);
+  inputMarkupEvents->InsertNextTuple1(vtkMRMLMarkupsNode::PointModifiedEvent);
+  inputMarkupEvents->InsertNextTuple1(vtkMRMLTransformableNode::TransformModifiedEvent);
+  vtkNew<vtkStringArray> inputMarkupClassNames;
+  inputMarkupClassNames->InsertNextValue("vtkMRMLMarkupsFiducialNode");
+  inputMarkupClassNames->InsertNextValue("vtkMRMLMarkupsLineNode");
+  inputMarkupClassNames->InsertNextValue("vtkMRMLMarkupsPlaneNode");
+  inputMarkupClassNames->InsertNextValue("vtkMRMLMarkupsAngleNode");
+  NodeInfo inputMarkups(
+    "Markups",
+    "Markups to specify spatial revolution axis. Normal for plane and angle, superior axis for a point.",
+    inputMarkupClassNames,
+    REVOLVE_INPUT_MARKUPS_REFERENCE_ROLE,
+    /*required*/ true,
+    /*repeatable*/ false,
+    inputMarkupEvents
+  );
+  this->InputNodeInfo.push_back(inputMarkups);
+
+  /////////
+  // Outputs
+  NodeInfo outputModel(
+    "Revolved model",
+    "Result of the revolving operation.",
+    inputModelClassNames,
+    REVOLVE_OUTPUT_MODEL_REFERENCE_ROLE,
+    false,
+    false
+  );
+  this->OutputNodeInfo.push_back(outputModel);
+
+  /////////
+  // Parameters
+
+  ParameterInfo parameterRotationAngleDegress(
+    "Rotation degrees",
+    "Rotation angle in degrees. Ignored for angle markup.",
+    REVOLVE_ANGLE_DEGREES,
+    PARAMETER_DOUBLE,
+    90.0);
+  this->InputParameterInfo.push_back(parameterRotationAngleDegress);
+
+  ParameterInfo parameterRotationAxisAtOrigin(
+    "Rotation axis at origin",
+    "If true, the revolution will be done around the origin.",
```

Comment:
Around the origin of what? Can you describe when this is useful?

Reply:
rigid transformations using a 4x4 matrix can be considered as a rotation around an axis that is not on the origin but elsewhere

Reply:
Can you describe some typical use cases?

Reply:
You may want to create a spring with an axis and center of rotation far from the world coordinates' origin